### ISSUE TITLE
getKeyDataFromString hash validation fail

### ISSUE THREAD

**Issue by xwiz** [paragonie/halite#166]
I've double-checked the hex file being loaded but still getKeyDataFromString always throws `Checksum validation fail`.

Is this some sort of OS encoding issue? What could be the issue?

**Comment by paragonie-security** [comment]
What do you mean by "the hex file being loaded"? Can you share a code snippet and a random throw-away example key that reproduces the same issue?

Typically when we see this error, someone thought they could just `bin2hex(random_bytes(32))` some random data and load it into Halite that way. 

(In reality, you're supposed to generate them within Halite, which produces a key in the form of `raw_key || checksum` before hex-encoding. This offers two advantages: It informs you when the key has been corrupted on disk, and it prevents people from easily using `hash('sha256', $password, true)` as their encryption key.)

**Comment by xwiz** [comment]
Interesting. So if I already have a key pair, how do I encode them to be 'halite compatible'?

**Comment by paragonie-security** [comment]
https://github.com/paragonie/halite/blob/7596d5cb25154092b524c34cb9ce2201db612ffc/src/KeyFactory.php#L769-L778

**Comment by xwiz** [comment]
OK thank you.

**Comment by xwiz** [comment]
I made a small helper as below and exported the new file to disk but it now says it's invalid _(must be CRYPTO_SIGN_PUBLICKEYBYTES long)_. I attached a sample public key output from the function.

What am I doing wrong?

	/**
     * Export a cryptography key file to halite compatibe file format (with a checksum)
     *
     * @param string $key_path
     * @param string $dest_path
     * @return HiddenString
     *
     * @throws CannotPerformOperation
     * @throws InvalidType
     * @throws \SodiumException
     * @throws \TypeError
     */
	public static function exportRawKey($key_path, $dest_path)
	{
		$rawKey = \file_get_contents($key_path);
		if ($rawKey === false) {
            // @codeCoverageIgnoreStart
            throw new \ParagonIE\Halite\Alerts\CannotPerformOperation(
                'Cannot load file: '. $key_path
            );
            // @codeCoverageIgnoreEnd
        }
		$key = new \ParagonIE\HiddenString\HiddenString(
			\ParagonIE\ConstantTime\Hex::encode(
				\ParagonIE\Halite\Halite::HALITE_VERSION_KEYS . $rawKey .
				\sodium_crypto_generichash(
					\ParagonIE\Halite\Halite::HALITE_VERSION_KEYS . $rawKey,
					'',
					\SODIUM_CRYPTO_GENERICHASH_BYTES_MAX
				)
			)
		);
		\file_put_contents($dest_path, $key->getString());
		return $key;
	}


[fst_rsa_halite.txt](https://github.com/paragonie/halite/files/6472994/fst_rsa_halite.txt)

**Comment by paragonie-security** [comment]
> fst_rsa_halite.txt

You cannot use RSA keys with Halite. Please generate a new keypair within Halite and use that instead.

**Comment by xwiz** [comment]
OK thank you. I am not familiar with the key format generated by sodium.

 However I noticed there's already a function to save/export in [KeyFactory class](https://github.com/paragonie/halite/blob/7596d5cb25154092b524c34cb9ce2201db612ffc/src/KeyFactory.php#L882)

Also the documentation to generate a signature keypair references the `SignatureKey` class instead of `KeyFactory`.

**Comment by paragonie-security** [comment]
Yep, the documentation for `KeyFactory` is available here: https://github.com/paragonie/halite/blob/master/doc/Classes/KeyFactory.md
